### PR TITLE
Enhancement: Truffle Test, ability to give multiple directories as an argument instead of only files.

### PR DESCRIPTION
**Pull request by swisstackle**
As of #5066 , I came up with the idea to allow one or multiple directories as an argument where all the files are being executed within those directories. It allows the sc-tester to organize the tests more thoroughly.

This is just a prototype and has to be thoroughly tested. Where do you guys locate unit tests of commands that do not have their own package in /truffle/packages? 

At the moment, the way it works is to recursively search for every file within the given directories. Then they are all added to filesToRun[].

Also, I am sure there is a lot to criticize about my code, just let me know if you see anything.

### CODE REVIEW

**Comment by swisstackle** [comment]
I figured out that the unit tests are located in packages/core/test/lib/commands/test.js.
I will add some tests there.

**Comment by haltman-at** [comment]
Oops, looks like nobody noticed that the CI failure was spurious and reran the tests until now!  Guess we'll have to find people to review this...

File: packages/core/lib/commands/test/determineTestFilesToRun.js [hunk 1]
```diff
@@ -1,27 +1,65 @@
...
+        filesToRun.push(fileOrDir);
+      }
+      // if the argument is a directory, run the recursive algorithm to detect every file within the directory and its' subdirectories.
+      else if(isDirectoryExists){
+        const files = getFilesRecursive(fileOrDir,[]);
+        files.forEach((file) => {
+          filesToRun.push(file);
+        });
+      }
+      else{
```
Comment (eggplantzzz) [inline packages/core/lib/commands/test/determineTestFilesToRun.js:25]: You can get rid of this `else` at the end.

File: packages/core/lib/commands/test/determineTestFilesToRun.js [hunk 2]
```diff
@@ -1,27 +1,65 @@
...
-  const path = require("path");
-  const fs = require("fs");
-  const glob = require("glob");
   let filesToRun = [];
-
   if (inputFile) {
     filesToRun.push(inputFile);
-  } else if (inputArgs.length > 0) {
-    inputArgs.forEach(inputArg => filesToRun.push(inputArg));
+  }else if (inputArgs.length > 0) {
```
Comment (eggplantzzz) [inline packages/core/lib/commands/test/determineTestFilesToRun.js:9]: Some Truffle style things; could you add spaces before and after your `else if`s like `} else if {` and not like `}else if{`?

Could you also add spaces in your `for` loop as well like `for (...) {` and not `for(...){`?

Additionally could you put your `else if`s on the same line as the closing bracket before it? So `} else if {` and not
```
}
else if {
```?

File: packages/core/lib/commands/test/determineTestFilesToRun.js [hunk 3]
```diff
@@ -1,27 +1,65 @@
...
+      let isFileExists = (fs.existsSync(fileOrDir) && fs.statSync(fileOrDir).isFile());
+      let isDirectoryExists = (fs.existsSync(fileOrDir) && fs.statSync(fileOrDir).isDirectory());
+      // If the argument is a file, add the file to filesToRun
+      if(isFileExists){
+        filesToRun.push(fileOrDir);
+      }
+      // if the argument is a directory, run the recursive algorithm to detect every file within the directory and its' subdirectories.
+      else if(isDirectoryExists){
+        const files = getFilesRecursive(fileOrDir,[]);
+        files.forEach((file) => {
```
Comment (eggplantzzz) [inline packages/core/lib/commands/test/determineTestFilesToRun.js:21]: Not necessary but if you want to be slick here do `files.forEach(file => filesToRun.push(file))`. It does return the thing you are adding but it shouldn't matter here.

File: packages/core/lib/commands/test/determineTestFilesToRun.js [hunk 4]
```diff
@@ -1,27 +1,65 @@
...
+/*
+ * @param {String} dirPath - A string that represents the relative or absolute path of the directory to search through.
+ * @returns {array[String]} - An array of all files in the directory dirPath, including all files of all subdirectories in dirPath.
+ */
+const getFilesRecursive = (dirPath, arrayOfFiles) => {
+  let files = fs.readdirSync(dirPath);
+  arrayOfFiles = arrayOfFiles || [];
+
+  files.forEach((file) => {
+    if (fs.statSync(dirPath + "/" + file).isDirectory()) {
```
Comment (cds-amal) [inline packages/core/lib/commands/test/determineTestFilesToRun.js:52]: File path segments should be concatenated with [path.join](https://nodejs.org/dist/latest-v18.x/docs/api/path.html#pathjoinpaths) which uses the platform-specific separator to delimit segments.

File: packages/core/lib/commands/test/determineTestFilesToRun.js [hunk 5]
```diff
@@ -1,27 +1,65 @@
...
   return filesToRun.filter(file => {
     return file.match(config.test_file_extension_regexp) !== null;
   });
 };
 
+/*
+ * @param {String} dirPath - A string that represents the relative or absolute path of the directory to search through.
+ * @returns {array[String]} - An array of all files in the directory dirPath, including all files of all subdirectories in dirPath.
+ */
+const getFilesRecursive = (dirPath, arrayOfFiles) => {
```
Comment (cds-amal) [inline packages/core/lib/commands/test/determineTestFilesToRun.js:47]: This function doesn't consider sym-linked folders. How would this function behave with the following structure below, where `a-prime` is a symbolic link to `a`?
```
a
├── a-prime -> ../a
└── a.txt
```

**Comment by swisstackle** [comment]
@cds-amal Do you mean things like hidden folders?

**Comment by cds-amal** [comment]
[This video is a super high-level overview/intro](https://www.youtube.com/watch?v=4-vye3QFTFo).

**Comment by swisstackle** [comment]
@cds-amal I believe that fs.readLinkSync(linkOrDir) just returns the original directory of the link. And if "linkOrDir" is not a symbolic link, it will just return the original linkOrDir.
That said, we can just use this function and then we don't have to use stats.isSymbolicLink.

**Comment by swisstackle** [comment]
I was wrong. I had to use isSymbolicLink regardless.

**Comment by benjamincburns** [comment]
I think this would likely be a lot cleaner if a module like [`walkdir`](https://www.npmjs.com/package/walkdir) was used. You may want to check the existing truffle packages to see if we already use that package, or something similar.

**Comment by swisstackle** [comment]
@benjamincburns Ok, will do that.

**Comment by eggplantzzz** [comment]
Hey @swisstackle, any updates on this? Or do you have any difficulties that you'd like to ask about for clarification?

**Comment by swisstackle** [comment]
Ready to test. I tested it locally with the following file structure:
test->test2->PlayersTestDir.js
test->thisislink2 (sysLink) -> ../test3/PlayerTestDir.js
test->thisislink (sysLink) -> test->test2->PlayersTestDir.js
PlayersTest.js

**Comment by swisstackle** [comment]
Just a sidenote: When running "yarn bootstrap" I'm getting windows  compatibility issues again:
@truffle/contract-schema: $ sh ./scripts/generate-declarations 
@truffle/contract-schema: 'sh' is not recognized as an internal or external command,

I made it work by using Gitbash, that recognizes the sh command.

Sidenote2: I am not sure if the yarn lock changes are meant to be as extensive as they are. I ran yarn lock after I added walkdir to the scope of packages/core and those are the changes it made to the yarn lock file.

File: packages/core/lib/commands/test/determineTestFilesToRun.js [hunk 6]
```diff
@@ -1,15 +1,23 @@
...
-  const glob = require("glob");
   let filesToRun = [];
-
   if (inputFile) {
     filesToRun.push(inputFile);
   } else if (inputArgs.length > 0) {
-    inputArgs.forEach(inputArg => filesToRun.push(inputArg));
+    for (let fileOrDir of inputArgs) {
+      fileOrDir = path.join(fileOrDir);
+      walkdir.sync(fileOrDir, { follow_symlinks: true }, function (path, stat) {
```
Comment (eggplantzzz) [inline packages/core/lib/commands/test/determineTestFilesToRun.js:13]: Can we omit the callback here? If we are going to use `walk.sync` I would kind of prefer to see something like
```
const results = walkdir.sync(fileOrDir, { follow_symlinks: true });
const isFile = fs.statSync...
// and so on
```

**Comment by eggplantzzz** [comment]
Thanks again @swisstackle! I took a look at walkdir and from my experiments it looks like it doesn't get caught in recursive loops in the case the symlink is linked to some parent folder. So I guess that is good. What do you think @cds-amal?

**Comment by swisstackle** [comment]
I believe that my comments are not going through and are stuck at pending (within the review). So here we go: "Ok. Will do. Are the files in "filesToRun" being checked later on if they are js files or not? If not, I should maybe check whether they end with .js"

File: packages/core/lib/commands/test/determineTestFilesToRun.js [hunk 7]
```diff
@@ -1,15 +1,23 @@
...
-  const fs = require("fs");
-  const glob = require("glob");
   let filesToRun = [];
-
   if (inputFile) {
     filesToRun.push(inputFile);
   } else if (inputArgs.length > 0) {
-    inputArgs.forEach(inputArg => filesToRun.push(inputArg));
+    for (let fileOrDir of inputArgs) {
+      fileOrDir = path.join(fileOrDir);
```
Comment (eggplantzzz) [inline packages/core/lib/commands/test/determineTestFilesToRun.js:12]: This line is probably unnecessary now that things have changed a bit, right?

**Comment by eggplantzzz** [comment]
Yeah it checks them [here](https://github.com/trufflesuite/truffle/blob/develop/packages/core/lib/commands/test/determineTestFilesToRun.js#L20-L22). They can be a variety of file types here, including Solidity.

File: packages/core/lib/commands/test/determineTestFilesToRun.js [hunk 8]
```diff
@@ -1,15 +1,23 @@
...
-  const path = require("path");
-  const fs = require("fs");
-  const glob = require("glob");
   let filesToRun = [];
-
   if (inputFile) {
     filesToRun.push(inputFile);
   } else if (inputArgs.length > 0) {
-    inputArgs.forEach(inputArg => filesToRun.push(inputArg));
+    for (let fileOrDir of inputArgs) {
```
Comment (cds-amal) [inline packages/core/lib/commands/test/determineTestFilesToRun.js:11]: `walkdir.sync` will throw if any member of `inputArgs` does not exist. `inputArgs` should be filtered to only include valid folders, possibly with some messaging to alert the user of the invalid entries.

Comment (cds-amal) [reply packages/core/lib/commands/test/determineTestFilesToRun.js:11]: Additionally, if a user specifies empty directories, the next if statement checking `filesToRun.length` will  pass and the default test directory suite will run. This is strange behavior if a user specifically did not mention the default `test` folder.

**Review by cds-amal — CHANGES_REQUESTED** [review]
Hi @swisstackle, I appreciate the progress you've made and think we're close to nailing the functionality. I'd like to see tests added to validate the expected behavior.
when arg is:
 - a non-existing directory, the user is notified and no runtime exceptions.
 - empty directories, the tests under default test folder should not be included
 - a folder structure with a cycle (via symlink), the user should be notified of the cycle and the  tests should run. Alternatively, if a cycle is detected, throw a `@truffle/error` with appropriate messaging

**Comment by swisstackle** [comment]
@eggplantzzz Do you see my comments inside the suggestion-thread? For some reason it says "pending" everytime I comment.

**Comment by benjamincburns** [comment]
> Do you see my comments inside the suggestion-thread? For some reason it says "pending" everytime I comment.

@swisstackle usually this happens when you've started a review. To publish the comments go to the "Files changed" view in this PR, and click the green "Review changes" button near the top right, select the "comment" radio button in the dialog that appears, and then click the "Submit review" button on the bottom corner of the dialog.

image

File: packages/core/lib/commands/test/determineTestFilesToRun.js (hunk 8 above)
Comment (cds-amal) [reply packages/core/lib/commands/test/determineTestFilesToRun.js:11]: @benjamincburns it should fail fast and quick in this case

**Comment by cds-amal** [comment]
Hey @swisstackle, 

I'd like to close this PR and open a new one. Please see my reasoning below.

Some tests against your branch revealed some behavior that seems incorrect. More importantly, I found it difficult to follow conversations, which might be due to using top-level comments or the review interface to respond to unresolved conversations. GitHub allows any team member to create a conversation via the review interface, as well as contribute to an existing conversation by responding to the comment. Please use the "respond to conversation" feature when appropriate. Of course, you can use the review interface to bring attention to a range of code, which creates a new conversation... or add a top-level comment if there's no need to reference code.

The feature you've suggested would improve the truffle test story, and think we can get to a proper implementation if we first outline the requirements to lock down expected behavior (in a new PR).

This new `truffle test` proposal alters the number of arguments to the command, and we should account for it. For example, what's the expected behavior for the following commands?

```console
truffle test 
truffle test test/metacoin.js
truffle test test/metacoin.js testFolder1 testFolder2...
truffle test test/a.js test/b.js test/c.js
```

> This is just a prototype and has to be thoroughly tested. Where do you guys locate unit tests of commands that do not have their own package in /truffle/packages? 

Truffle has unit tests that live in the same `packages/project` folder and integration tests in the `packages/truffle`  folder.

> Also, I am sure there is a lot to criticize about my code, just let me know if you see anything.

I'm glad you're open to feedback and hope we can push this across the finish line, and  hope my comments make sense. If you have questions, please ask! So, what

## metacoin tests

I set up test project based on the metacoin-box (`truffle unbox metacoin mc`) with two test folders
 - test - has js tests
 - test-solidity - has solidity tests


<details><summary>Test layout</summary>
<p>

```console
╭─amal@falafel  ~/.scratch/2022/07/18/swiss-dirs/mc ‹node-16.16.0›  ‹›
╰─$ tree
.
├── contracts
│   ├── ConvertLib.sol
│   ├── MetaCoin.sol
│   └── Migrations.sol
├── LICENSE
├── migrations
│   ├── 1_initial_migration.js
│   └── 2_deploy_contracts.js
├── test
│   └── metacoin.js
├── test-solidity
│   └── TestMetaCoin.sol
└── truffle-config.js

4 directories, 9 files

```
</p>
</details>


## truffle test

```console
╭─amal@falafel  ~/.scratch/2022/07/18/swiss-dirs/mc ‹node-16.16.0›  ‹›
╰─$ truffle test
```
<details><summary>results as expected</summary>
<p>

```console
Compiling your contracts...
===========================
> Compiling ./contracts/ConvertLib.sol
> Compiling ./contracts/MetaCoin.sol
> Compiling ./contracts/Migrations.sol
> Artifacts written to /tmp/test--1743274-Y2q2VdU15dGG
> Compiled successfully using:
   - solc: 0.5.16+commit.9c3226ce.Emscripten.clang


  Contract: MetaCoin
    ✔ should put 10000 MetaCoin in the first account
    ✔ should call a function that depends on a linked library
    ✔ should send coin correctly (1077ms)


  3 passing (1s)
```

</p>
</details>


## truffle test path/to/file

```console
╭─amal@falafel  ~/.scratch/2022/07/18/swiss-dirs/mc ‹node-16.16.0›  ‹›
╰─$ truffle test test/metacoin.js
```

<details><summary>results as expected</summary>
<p>

```console
Compiling your contracts...
===========================
> Compiling ./contracts/ConvertLib.sol
> Compiling ./contracts/MetaCoin.sol
> Compiling ./contracts/Migrations.sol
> Artifacts written to /tmp/test--1743388-0nsXaTr7XjbR
> Compiled successfully using:
   - solc: 0.5.16+commit.9c3226ce.Emscripten.clang


  Contract: MetaCoin
    ✔ should put 10000 MetaCoin in the first account
    ✔ should call a function that depends on a linked library
    ✔ should send coin correctly (1063ms)


  3 passing (1s)
```
</p>
</details>


## truffle test /path/to/folder

```console
╭─amal@falafel  ~/.scratch/2022/07/18/swiss-dirs/mc ‹node-16.16.0›  ‹›
╰─$ truffle test test-solidity
```

<details><summary>results as expected</summary>
<p>

```console

Compiling your contracts...
===========================
> Compiling ./contracts/ConvertLib.sol
> Compiling ./contracts/MetaCoin.sol
> Compiling ./contracts/Migrations.sol
> Compiling ./test-solidity/TestMetaCoin.sol
> Artifacts written to /tmp/test--1743534-BBNscOqTSZk2
> Compiled successfully using:
   - solc: 0.5.16+commit.9c3226ce.Emscripten.clang


  TestMetaCoin
    ✔ testInitialBalanceUsingDeployedContract (1056ms)
    ✔ testInitialBalanceWithNewMetaCoin (1040ms)


  2 passing (20s)
```

</p>
</details>

## truffle test /path/to/empty/directory

```console
╭─amal@falafel  ~/.scratch/2022/07/18/swiss-dirs/mc ‹node-16.16.0›  ‹›
╰─$ mkdir test-empty
╭─amal@falafel  ~/.scratch/2022/07/18/swiss-dirs/mc ‹node-16.16.0›  ‹›
╰─$ truffle test test-empty
```

<details><summary>Whoa, runs unspecified tests</summary>
<p>


```console
Compiling your contracts...
===========================
> Compiling ./contracts/ConvertLib.sol
> Compiling ./contracts/MetaCoin.sol
> Compiling ./contracts/Migrations.sol
> Artifacts written to /tmp/test--1744145-aVxVEeMoq6AW
> Compiled successfully using:
   - solc: 0.5.16+commit.9c3226ce.Emscripten.clang


  Contract: MetaCoin
    ✔ should put 10000 MetaCoin in the first account
    ✔ should call a function that depends on a linked library
    ✔ should send coin correctly (1073ms)


  3 passing (1s)

```
</p>
</details>


## Unclear API: truffle test /path/to/file /path/to/dir/with/tests

What does it mean when truffle test is invoked with a file and a folder?

```console
╭─amal@falafel  ~/.scratch/2022/07/18/swiss-dirs/mc ‹node-16.16.0› ‹ruby-3.0.0› ‹›
╰─$ truffle test test/metacoin.js test-solidity
```

<details><summary>???</summary>
<p>


```console
Compiling your contracts...
===========================
> Compiling ./contracts/ConvertLib.sol
> Compiling ./contracts/MetaCoin.sol
> Compiling ./contracts/Migrations.sol
> Compiling ./test-solidity/TestMetaCoin.sol
> Artifacts written to /tmp/test--1749328-NBlMKvd6ADTu
> Compiled successfully using:
   - solc: 0.5.16+commit.9c3226ce.Emscripten.clang


  TestMetaCoin
    ✔ testInitialBalanceUsingDeployedContract (1058ms)
    ✔ testInitialBalanceWithNewMetaCoin (1045ms)


  2 passing (20s)

```
</p>
</details>

**Comment by cds-amal** [comment]
Hey @swisstackle I just re-read my previous comment and want to make sure you know I am excited to collaborate to get your suggested feature added to Truffle. My motivation is based on the realization that this work needs more coordination than our team expected up-front. Apologies for that. Thanks again, and I look forward to working with you on this. :bow:

**Comment by swisstackle** [comment]
@cds-amal Dont worry, I fully understand. The problem I have is that when I chose to reply instead of opening a review, the comment stays pending (yellow pending symbol). I don't exactly know if this is a permissions problem or somethng else. We will see if it works in the new PR we will open up.

That being said, do you want me to open up a PR with the current code, to then make a more detailed plan? I am also open to hop on a call if you guys want to whenever you guys find time.

Until July 27th, I am usually available in the afternoon (incl saturday and sunday).

From July 29 - August 2nd I am available all day.

From August 3rd - August 31st I will not be available due to training camp (6am-9pm).

File: packages/core/lib/commands/test/determineTestFilesToRun.js (hunk 4 above)
Comment (swisstackle) [reply packages/core/lib/commands/test/determineTestFilesToRun.js]: test

File: packages/core/lib/commands/test/determineTestFilesToRun.js (hunk 8 above)
Comment (swisstackle) [reply packages/core/lib/commands/test/determineTestFilesToRun.js:11]: test

**Comment by swisstackle** [comment]
Let's close this. I am working on it locally with @cds-amal .

**Pull request closed without merging**